transport/awslambda/wrapper.go

```diff
@@ -0,0 +1,285 @@
+package awslambda
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+)
+
+// DecodeRequestWrapper wraps the given decoderSymbol function, and generates
+// the proper DecodeRequestFunc, based on the decoderSymbol function signature.
+// The decoderSymbol function signature has to receive 2 args, which is the
+// context.Context and event request to decode.
+// It has also to return 2 values, the user-domain object and error.
+func DecodeRequestWrapper(decoderSymbol interface{}) DecodeRequestFunc {
```

Comment:
I understand _what_ the functions in wrapper.go do, but I don't understand _why_ they're useful. Can you provide some more context?

Reply:
The motivation is to bring the `Decode` function purpose closer as decoder to user-domain object.

Without the wrapper function, in some use cases, the user actually needs to decode for 2 purposes, first into the AWS corresponding events, then next they decode it further to their business domain. 

The idea actually coming from the `lambda.Start()` in [aws-lambda-go](https://github.com/aws/aws-lambda-go/blob/master/lambda/entry.go). With this convenient function it become so simple to implement a lambda handler.

```
func show(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
...
}

func main() {
    lambda.Start(show)
}
```

With the wrapper it could help the consumer to focus on converting their user-domain from the familiar AWS events package.
```
// DecodeUppercaseRequest is a decoder of the JSON body into upperRequest.
func DecodeUppercaseRequest(
    _ context.Context,
    apigwReq events.APIGatewayProxyRequest,
) (
    UppercaseRequest,
    error,
) {
    request := UppercaseRequest{}
    err := json.Unmarshal([]byte(apigwReq.Body), &request)
    return request, err
}
```
and start the server like this
```
uppercaseHandler := awslambdatransport.NewServer(
    stringsvc.MakeUppercaseEndpoint(svc),
    awslambdatransport.DecodeRequestWrapper(stringsvc.DecodeUppercaseRequest),
    awslambdatransport.EncodeResponseWrapper(stringsvc.EncodeResponse),
    awslambdatransport.ServerErrorEncoder(
        awslambdatransport.ErrorEncoderWrapper(stringsvc.EncodeError),
    ),
)
```

Reply:
What is the use case for _not_ using this wrapper?

Reply:
Use case for not using wrapper:
1. if performance matters more than convenient. As the wrapper is using reflections.
2. Provide the most possible open for extension. For example if we want to provide custom message format into Lambda.

Reply:
OK, thanks for the clarification. I'm not OK with a function in the public API that takes an empty interface, which must be one of a pre-ordained number of function types. I understand this is how the AWS lambda.Start function works, but that doesn't excuse the bad practice.

I'd like to see this whole layer removed. The most obvious way I can think of is to provide separate, explicit wrapper functions for each of the supported function signatures. If there were a way to remove the need for this altogether, I'd prefer that, but I can't really judge if there's an acceptable way to do that.

Reply:
I see your point, yes you are right with API that takes empty interface{} its actually not trivial to understand its usage just from looking at the package itself, only if one see a whole lot of example of usages which means the package cant stand by its own.

OK - removing the wrapper. Thanks @peterbourgon  !

Reply:
updated 5a7cbf48baf113efd6ec7c15a9afecc13d7f00f5